### DISCUSSION TITLE
How to specify git dependencies with conditional Python version in `pyproject.toml`?

### DISCUSSION THREAD

**Discussion by EwoutH** [pypa/setuptools#3651] [Q&A] [answered]
I'm trying to install a package (statsmodels in this case) from git if on Python 3.11 or higher, and otherwise reguarly from PyPI. I would expect I have to specify it like this:
```toml
[project]
dependencies = [
    "numpy",
    "pandas",
    "statsmodels; python_version < '3.11'",
    "statsmodels @ git+https://github.com/statsmodels/statsmodels.git; python_version >= '3.11'",
]
```
But that results in the [following error](https://github.com/quaquel/EMAworkbench/actions/runs/3340064946/jobs/5529746737):
```
      configuration error: `project.dependencies[{data__dependencies_x}]` must be pep508
      DESCRIPTION:
          Project dependency specification according to PEP 508
      
      GIVEN VALUE:
          "statsmodels @ git+https://github.com/statsmodels/statsmodels.git; python_version >= '3.11'"
      
      OFFENDING RULE: 'format'

ValueError: invalid pyproject.toml config: `project.dependencies[{data__dependencies_x}]`.
configuration error: `project.dependencies[{data__dependencies_x}]` must be pep508
        "$id": "#/definitions/dependency",
        "title": "Dependency",
        "type": "string",
        "format": "pep508"
    }
```
Conditional Python version works:
```toml
[project]
dependencies = [
    "numpy",
    "pandas",
    "statsmodels; python_version < '3.11'",
]
```
Installing from git also:
```toml
[project]
dependencies = [
    "numpy",
    "pandas",
    "statsmodels @ git+https://github.com/statsmodels/statsmodels.git",
]
```
But when combined, setuptools can't parse is correctly.

Do I need to specify it differently? Or is this a bug?

(setuptools 65.5.0, and happens on all Python versions)

**Comment by jwodder** [comment, marked as answer]
I believe you simply need to insert a space between the URL and the semicolon; otherwise, the semicolon will be interpreted as part of the URL, resulting in a parse error.

**Reply by EwoutH** [reply]
That works, thanks, you're a hero!

**Reply by EwoutH** [reply]
I created a feature request to interpreted a semicolon as the end of an URL, or add an improved warning message: https://github.com/pypa/setuptools/issues/3652

**Comment by avalonv** [comment]
Recently ran into this issue too. Would be lovely if this was more clearly documented.
